### ISSUE TITLE
`nan` issues

### ISSUE THREAD

**Issue by yebai** [chalk-lab/Mooncake.jl#807]
See, https://github.com/chalk-lab/Mooncake.jl/commit/b8c538b0ee084b8ce1ced66d897f7d227d3a74a4#commitcomment-167818443

**Comment by AstitvaAggarwal** [comment]
@yebai @Technici4n , Note that in Forward Mode AD the issue of tangent poisoning due to singularies, `NaN` does not occur due to independent tangent propagation for each input (unlike Reverse Mode) . But if the mathematically correct tangent is something `Inf/NaN` it could still cause a mathematically expected fail in downstream use despite being no direct fault/improper tangent propagation within the AD library per se.

Just noticed that `ForwardDiff.jl` in non `nan_safe` mode returns a mathematically incorrect tangent. It should have returned `NaN` correctly like in `Mooncake.jl`.
```
julia> function f(x)
           a = x
           b = sqrt(abs(x-1.0))
           return b
       end

julia> DifferentiationInterface.derivative(f, AutoForwardDiff(), 1.0)
Inf

julia> DifferentiationInterface.derivative(f, AutoMooncakeForward(), 1.0)
NaN
```

But yeah I think if possible, we could add something similar to `nan_safe` mode (depends on Preferences.jl) for Mooncake as well. This would allow in Forward/Reverse Mode, returning **both** mathematically correct results  **and** something much more safe for ML, optimization routines, depending on what the user wants to work with.  I think JAX also does something similar...

**Comment by Technici4n** [comment]
I don't understand the justification for why forward and reverse mode should give different results. For a sufficiently complex function, it seems to me that a NaN in the middle of the computational graph will poison tangents, regardless of whether you are going forward or in reverse.

**Comment by AstitvaAggarwal** [comment]
This example in the [commit's](https://github.com/chalk-lab/Mooncake.jl/issues/807#issue-3511658870) discussion is a bit misleading and threw me off,
```
julia> function f(x)
           a = x
           b = sqrt(abs(x-1.0))
           return b
       end

julia> DifferentiationInterface.derivative(f, AutoForwardDiff(), 1.0)
Inf

julia> DifferentiationInterface.derivative(f, AutoMooncakeForward(), 1.0)
NaN

julia> function f(x)
           a = x
           b = sqrt(abs(x-1.0))
           return a
       end
f (generic function with 1 method)

julia> DifferentiationInterface.derivative(f, AutoMooncakeForward(), 1.0)
1.0

julia> DifferentiationInterface.derivative(f, AutoForwardDiff(), 1.0)
1.0
```
The `NaN` for `f`'s derivative (when we return b) wrt input `x` is just the correct undefined derivative at that point, not poisoning. And we also get correct derivative for when we return `a`. Here the `NaN` is just the correct undefined derivative, so returning either `a` or `b` gives mathematically correct tangents. (unlike prev [issue's MWE](https://github.com/chalk-lab/Mooncake.jl/issues/794#issue-3473595950)).

But yeah @Technici4n, poisoning actually does happen albeit very differently - referring to ForwardDiff's [docs](https://juliadiff.org/ForwardDiff.jl/v0.7/user/advanced.html#Fixing-NaN/Inf-Issues-1).
Will open up a PR for this, thanks!

**Comment by Technici4n** [comment]
Since there are conflicting goals (mathematical correctness VS avoiding poisoning, and potential consistency between forward and reverse mode), it would be good to think a bit more about what should happen at non-differentiable points. Maybe the result of the discussion should be formalized in the docs.

It seems that we (DFTK) encountered a poisoning issue in forward mode as well, and implemented a similar workaround, see https://github.com/JuliaMolSim/DFTK.jl/blob/38b7f27778ac2f2b7a92e6dbe4834c7e123a33d8/src/workarounds/forwarddiff_rules.jl#L310.

**Comment by Technici4n** [comment]
Related discussion: https://juliadiff.org/ChainRulesCore.jl/dev/maths/nondiff_points.html

**Comment by yebai** [comment]
It would be helpful to have a debug mode (eg, `AutoMooncake(error_on_nan_or_inf=true`) that errors loudly on `inf` / `nan` tangents, so one can easily identify the source of these numerical issues during optimisation / MCMC.

**Comment by Technici4n** [comment]
Note that PyTorch will return Inf as the gradient of sqrt(x):
```py
import torch
from torch.func import grad
grad(lambda x: torch.sqrt(x))(torch.tensor(0.0))
```

It is also happy to return NaN in the VJP:
```py
import torch
from torch.func import vjp
vjp(lambda x: torch.sqrt(x), 0.0)[1](torch.tensor(0.0))
# -> (tensor(nan),)
```

However if I give it the original example from #794, it does correctly compute the gradient:
```py
import torch
from torch.func import vjp
def f(x):
    a = x
    b = torch.sqrt(torch.abs(x - 1.0))
    return a

vjp(f, torch.tensor(1.0))[1](torch.tensor(1.0))
# -> (tensor(1.),)
```

I suppose the conclusion is that returning Inf/NaN is preferable, and that Mooncake should be fixed to avoid poisoning from "inactive variables".

**Comment by yebai** [comment]
Thanks @Technici4n for investigating these. 

Would you like to take the lead on this and formulate a practical, concrete plan to address it?

**Comment by Technici4n** [comment]
I am happy to continue brainstorming, not sure I can take the lead. It would be helpful if developers of other AD systems / rules would give their perspective.

Considering only `sqrt(x)`, I suppose that there are three solutions for the issue in #794:
1. View `sqrt(x)` as
    ```math
    \begin{cases} 0, & x = 0, \\ \sqrt x, & x > 0, \end{cases}
    ```
    and then take the derivative considering all decision branches frozen. Effectively this means that `sqrt'(0) = 0`. This is what was implemented for rrules already in https://github.com/chalk-lab/Mooncake.jl/commit/b8c538b0ee084b8ce1ced66d897f7d227d3a74a4, and if chosen should be extended to forward mode as well. This is also what Enzyme is doing.
2. Set `sqrt'(0) = Inf`, but implement "NaN-safe" propagation, such that if the tangent is 0 the derivative gets propagated as 0 rather than a NaN. This is what ForwardDiff does if NaN-safe mode is enabled.
3. Set `sqrt'(0) = Inf`, don't implement NaN-safe propagation, and perform enough dead code elimination in the Mooncake compiler to avoid poisoning the gradient like in #794. This might unfortunately be "impossible" in the presence of mutation and/or aliasing.

(As a side-note, Mooncake already defines `abs'(0) = 0`.)

3 is for sure the most elegant solution, but might be impossible. Unless a way to mark variables as "unused" is added? (Sounds a lot like Enzyme rules, doesn't it? 😄) Otherwise my preferred option would I suppose be 1.

**Comment by yebai** [comment]
Perhaps we can define a new number type `NotAGradient <: Real` (or `NaG` as a shorthand) and use it instead of `Nan` or `Inf`? 

This would easily allow us to avoid both gradient poisoning and returning mathematically incorrect gradients (e.g., `0`).

**Comment by Technici4n** [comment]
Sounds like a good idea, similar to JAX's `ad.Zero` I suppose. That is basically what I was getting at with marking variables as "unused". It would mean that rrules need to be aware of it, so it would be a large design change. @willtebbutt thoughts?

**Comment by gdalle** [comment]
That's probably not a good idea because it would introduce type instabilities. The type of `NaN` is the same as that of the surrounding numbers, it's just a special case of `Float64`, which is what makes it efficient.

**Comment by Technici4n** [comment]
Some activity analysis should hopefully fix this.

**Comment by yebai** [comment]
Looping back to the `NotAGradient <: Real` idea, a minor performance penalty is acceptable when traded for handling `NaN`. Additionally, `NotAGradient` could be offered as a debug option rather than being enabled by default.

**Comment by gdalle** [comment]
I think the performance penalty would be far from minor. For instance it would prevent subsequent functions from hitting BLAS dispatches

**Comment by AstitvaAggarwal** [comment]
The above issue is to handled by a PR performing changes as suggested in #938's docstrings - by changing the rules for *all* primitive functions who's adjoint becomes singular at certain points.

A Documentation page is to be added as suggested in thread https://github.com/chalk-lab/Mooncake.jl/pull/938#discussion_r2698827893 .
1. This would make it clear to users on handling this issue via non-primitive rules for functions with derived rules.
2. Cover any  "`NaN` handled" - missed out primitive rules that may cause potential future problems.
3. Improves Docs about handling edge cases etc.

**Comment by yebai** [comment]
To clarify, PR #938 only imports ChainRules rules for `SpecialFunctions`
functions whose arguments have `NotImplemented` tangents.

The general `NaN`-poisoning issue remains unresolved. Addressing it would
require intermediate representation (IR) analysis, which is outside the
scope of Mooncake.

As a result, Mooncake errs on the conservative side: it always returns
correct gradients when possible, but may occasionally return `NaN` in cases
where a more precise IR analysis capable of eliminating dead code
(i.e. code that does not contribute to gradients) could avoid it.

**Comment by AstitvaAggarwal** [comment]
Actually yes, but I meant the same approach around detecting if `dx==0` or not. And instead of `NaN`ifying the tangent we just zero it when it's unused.

**Comment by yebai** [comment]
**Update**

Some low-level Mooncake rules implement NaN-safe propagation (i.e., `nan_tangent_guard`, also known as *strong zero* in ForwardDiff; Bruno’s option 2, see [here](https://github.com/chalk-lab/Mooncake.jl/issues/807#issuecomment-3416467289)). This approach eliminates NaN propagation from inactive code (i.e., code that does not contribute to gradients) without requiring IR analysis, provided that `nan_tangent_guard` is applied consistently across rules.

In practice, however, `nan_tangent_guard` is applied selectively. As a result, some primitives may still propagate NaNs even when they are part of inactive code. Please open an issue if such a case is discovered, or submit a PR to fix the relevant rule.

Note that NaNs arising from genuine numerical issues will still propagate through primitives that are part of active code.
